superchain-backend.md

```diff
@@ -0,0 +1,457 @@
+the safety level label just moves along that log, like it does on the chain of blocks.
+
+### Simple interface:
+
+The verifier side is really all just one interface:
+
+```
+type MessageVerifier interface {
+    Check(Identifier, PayloadHash) SafetyLevel
+}
+```
+
+This can be backed by:
+
+- Wrapper around an ethereum RPC:
+    - using `eth_getLogs`, fetch log with 1-block 1-contract narrow scope, check if it matches,
+      return safety-level of block it was included in.
+    - using `eth_getTransactionReceipt` (if tx hash is known), fetch logs of specific tx, rely on metadata for event-index,
+      return safety-level of block it was included in.
+    - using `eth_getBlockReceipts`, fetch all logs of block, check the specific log,
+      return safety-level of block it was included in.
+- Wrapper around a append-log "database", as described in the ideas above:
+  search the database for the data, check against computed message-hash, and then return safety level.
+- Wrapper around a shreds reader: if we maintain the tip of the chain in memory based on streamed data,
+  we can serve queries for that data, and otherwise fallback to another source.
+
+In a fault-proof, a trustless version of the `eth_getBlockReceipts` can support the query,
+as that verifies the log-index correctness relative to the start of the block, and inclusion-proof to the block etc.
+
+### Batching
+
+We may want to consider batch-queries as part of the interface;
+query `[]BatchElem` (`BatchElem = (Identifier, PayloadHash)`), return `[]SafetyLevel`.
+
+L2 Blocks already form a natural form of batching, this should be supported by the interface also,
+but may not be applicable for per-message systems like the tx-pool.
+
+## Fault proving
+
+Note that chain-procesing is optimistic and asynchronous: chains are individually processed,
+and cross-L2 messaging safety is deferred to an asynchronous process.
+
+To support this in fault-proving, the "loop" of block-processing has to be enhanced with this cross-L2 verification.
+To avoid reorgs in the fault-proof, the block-processing and cross-L2 verification should be interleaved,
+to run one step at a time each.
+
+With an additional simplification of not allowing equivocation of batch-data,
+and falling back to a safe deposits-only L2 block, the proof is simplified to running one block of each
+L2 and then performing an aggregate cross-L2 verification step of this block-height.
+
+The "backend" here is backed by pre-image oracle queries that:
+- inspect the emitted tentative event-log hashes of the latest block-height, to extract executing messages
+- inspect the matching event-logs of the historical data, to retrieve initiating messages (if valid)
+
+
+# Alternatives Considered
+
+<!-- *List out a short summary of each possible solution that was considered.
+Comparing the effort of each solution* --> 
+
+## Original superchain-backend draft, with recursive resolution, RPC backing
```

Comment:
Maybe i'm misinterpreting but the original draft didn't trigger messages resolution recursively. Resolution was only triggered as initiating data that a block was dependent on has become available

Reply:
It's been a while, I have lost some of the context on these drafts. But I believe it was 1-deep resolution, and then relied on the same system resolving itself the same way, thus effectively recursive.